logstore: don't load LastIndex in Load{Term,Entries}

pav-kv opened the pull request:
This PR broadens the meaning of `ErrUnavailable` to be a generic "entry not found" kind of error, and partially removes excessive log boundary checks in `LoadTerm` and `LoadEntries`.

Most callers check the log boundaries before reading from the log, so they can deduce that `ErrUnavailable` means a gap in the log if needed, and can error/panic accordingly. For example, any misaligned call to log storage `slice()` in `raft` will [panic](https://github.com/cockroachdb/cockroach/blob/19290ae07afa1a296fca9eee9f77d2e9c60eb804/pkg/raft/log.go#L597-L601) (and would have panicked before this PR).

Callers above `raft` don't ever check for `ErrUnavailable`, and treat it as any other error, so this change is a no-op w.r.t. them.

Prereq for #142013
Related to #132114, #136109

pav-kv committed:
logstore: don't load LastIndex in LoadTerm

See the function comment. RawNode never accesses log indices above the
storage.LastIndex() because it has them covered by the unstable log.

Epic: none
Release note: none

pav-kv committed:
logstore: don't load LastIndex in LoadEntries

Epic: none
Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/142014)
<!-- Reviewable:end -->

pav-kv commented on pkg/kv/kvserver/logstore/logstore.go:
```diff
@@ -673,6 +673,14 @@ func LoadTerm(
 // does not exceed maxSize, unless only one entry is returned.
 //
+// There are 3 cases for when an entry is not found: (1) the lo index has been
+// compacted away, (2) the hi index is higher than LastIndex, or (3) there is a
```
`hi+1`

tbg replied:
"greater than or equal to"? I'm never quite sure if "higher" is exclusive or inclusive

tbg replied:
while you're here, the range is (lo,hi] (assuming everything got returned, right)? Maybe worth putting this in so that it's clear from this comment that lo is exclusive and hi inclusive.

pav-kv replied:
Made it ">" to be clear, and updated the comment to show lo/hi inclusivity. It's [lo, hi) rather than (lo, hi], though [I think](https://github.com/cockroachdb/cockroach/blob/c44b3134d23d6037e1d5d5e12461a6914aa588e3/pkg/raft/storage.go#L64-L66) it would be cleaner the other way around (and seen benefit of it in other raft code: lots of +-1 just disappear from the code/comments).

Re "higher": it means "greater", intuitively. Can't say that something is "higher" if it's level/equal.
<!-- Reviewable comment -OK16CqU5Sk0U2qm2IsD:b2tp3q -->
<!-- Sent from Reviewable.io -->

pav-kv commented:
TODO:
- [ ] update `raft.Storage` comments
- [ ] double-check what happens with all `term()` callers in `raft` if we return `ErrUnavailable` for a gap (whereas previously we would have returned an arbitrary error)

tbg reviewed (commented):
> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142014)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @hakuuww)

___
*[`pkg/kv/kvserver/logstore/logstore.go` line 605 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142014#-OK0Md2EBPxTwtuTgC7H:-OK0Md2EBPxTwtuTgC7I:biius47) ([raw file](https://github.com/cockroachdb/cockroach/blob/99ca279f477de47d2939a8246f4cb05b1b417f9a/pkg/kv/kvserver/logstore/logstore.go#L605)):*
> ```Go
> // range. The result is loaded from the storage engine if it's not in the cache.
> //
> // There are 3 cases for when the term is not found: (1) the index has been
> ```

The lowest possible index to pass in and not get an error is the index for the truncated state, right? i.e. we can go one lower than the first index in the log.
I see this is true from the code below, maybe worth mentioning here as well ("valid range is `[TruncatedState.Index,LastIndex]`")


<!-- Sent from Reviewable.io -->

tbg reviewed (commented):
Quick pass. Looks good to me - I note the TODOs, happy to take another closer look when you think this is fully ready.

pav-kv reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142014)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @hakuuww and @tbg)

___
*[`pkg/kv/kvserver/logstore/logstore.go` line 605 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/142014#-OK0Md2EBPxTwtuTgC7H:-OK15iSUCbWWfQiRP9Fr:b-9fne85) ([raw file](https://github.com/cockroachdb/cockroach/blob/99ca279f477de47d2939a8246f4cb05b1b417f9a/pkg/kv/kvserver/logstore/logstore.go#L605)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

The lowest possible index to pass in and not get an error is the index for the truncated state, right? i.e. we can go one lower than the first index in the log.
I see this is true from the code below, maybe worth mentioning here as well ("valid range is `[TruncatedState.Index,LastIndex]`")
</blockquote></details>

Done.

Yes, we can go 1 below the first index. NB: "first index" is a weird concept because it's undefined for an empty / fully compacted log; or, rather "weirdly defined" and points to a non-existing entry. So referring to "truncated index" sounds like a better idea. But it's an [exercise](https://github.com/cockroachdb/cockroach/blob/c44b3134d23d6037e1d5d5e12461a6914aa588e3/pkg/raft/storage.go#L91-L92) for some other time :)


<!-- Sent from Reviewable.io -->

hakuuww reviewed (approved):
Just to clarify, a gap in the log means there is a skip in raft entry index in the continuous raftLog? And I think this should never happen according to the raft algorithm. So something seriously wrong must have happened.

pav-kv committed:
raft: simplify lastEntryID

Epic: none
Release note: none

pav-kv committed:
raft: make raftLog.term() checks stronger

This commit makes raftLog/LogSnapshot panic when it detects a gap in the
log. The observation is that we should never expect ErrUnavailable from
storage for the part of the log that is supposed to be contiguous.

Epic: none
Release note: none

pav-kv commented:
@tbg Ready for another look. Done the TODOs (last two commits).

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/142014#-:-OKBIrCQ0-4b3AciMomE:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @pav-kv)



<!-- Sent from Reviewable.io -->

pav-kv commented:
TFTRs!

bors r=tbg,hakuuww
